### DISCUSSION TITLE
Adding new data augmentations

### DISCUSSION THREAD

**Discussion by astrobdr** [MIC-DKFZ/nnUNet#701] [Q&A]
Looking for general advice to add (some fairly complex) data augmentations to nnUNet.
If there is already documentation for doing this please point me to it, but I only found one line in `documentation/extending_nnunet.md` which said to look in `nnunet.training.network_training.data_augmentation.*`. That points me to the network trainers, as there is no `data_augmentation/` folder or file (at least not anymore). My assumption was that our data augmentations to the training data must happen before nnUNet is even aware of our network trainer.

Clearly, im very confused.

For some clarification, I am working on the kits21 data, and we already have a baseline nnUNet working with nnUNetTrainerV2. So I think I should modify the nnUNetTrainerV2 file?

Some questions:
1. should my augmentation happen before or after running `nnUNet_plan_and_preprocess`? Is `nnUNet_plan_and_preprocess` even aware of which network trainer we use?
2. Is there a way of testing these in an interactive environment such as jupyter notebook? I will be building fairly complex augmentations.

Even if you cant answer those questions, just giving me a direction to go would be a big help, thanks!

**Comment by FabianIsensee** [comment]
Augmentation should never be done offline. This just wastes disk space. You can do this on the fly during training. Thus, the trainer must be aware of it and implement it. The configuration of the data augmentation can depend on the plans and is thus part of the rule-based system nnU-Net is based on.
Augmentation is set in https://github.com/MIC-DKFZ/nnUNet/blob/fd58e25a304e2ab0cd4c16d2c79505c2882e3593/nnunet/training/network_training/nnUNetTrainerV2.py#L55. You should be able to follow the code from there.
Best,
Fabian
